### PR TITLE
Payment Tolerance

### PR DESCRIPTION
**Pull request by Kukks**
This PR is not tested yet. 
This adds a new configurable feature to allow stores to accept invoices which are under paid to a specified percentage. The default is 0% so no store marks invoices as paid if not fully paid unless configured. 

@NicolasDorier  Currently I've decided to:

- send the paid event with the name `invoice_paidWithinTolerance`
- set the exceptionstatus to  `paidWithinTolerance`
- set the status to `paid`

I'm unsure if `invoice_paidWithinTolerance` and `paidWithinTolerance` breaks any bitpay compatibility, I'm still new to this api. :)

### CODE REVIEW

File: BTCPayServer/Models/InvoiceResponse.cs [hunk 1]
```diff
@@ -178,7 +178,7 @@ public decimal Rate
         }
 
         //"exceptionStatus":false
-        //Can be `paidPartial`, `paidOver`, or false
+        //Can be `paidPartial`, `paidOver`, `paidWithinTolerance` or false
```
Comment (NicolasDorier) [inline BTCPayServer/Models/InvoiceResponse.cs]: for compatibility purpose I prefer we don't add new stuff here

**Comment by NicolasDorier** [comment]
I think it can be simplified.

I don't think we need event nor new exception status. (Especially the exceptionStatus which can break compatibility with bitpay)

The best is to just replace all `accounting.Paid < accounting.TotalDue` with `!accounting.Paid(tolerance)`.

**Comment by Kukks** [comment]
Did the changes @NicolasDorier

**Comment by NicolasDorier** [comment]
Actually you can simplify more: Use the `Money.Almost` methods ;)

File: BTCPayServer/Services/Invoices/InvoiceEntity.cs [hunk 2]
```diff
@@ -523,6 +524,22 @@ public class PaymentMethodAccounting
         /// Total amount of network fee to pay to the invoice
         /// </summary>
         public Money NetworkFee { get; set; }
+
+        public bool IsPaid(double tolerance, out bool paidOver)
+        {
+            paidOver = false;
+            if (Paid < TotalDue)
```
Comment (NicolasDorier) [inline BTCPayServer/Services/Invoices/InvoiceEntity.cs]: This whole method can be summarized as following:

```
var totalDue = TotalDue - (TotalDue * (tolerance/100));
return Paid <= totalDue;
```

Comment (NicolasDorier) [reply BTCPayServer/Services/Invoices/InvoiceEntity.cs]: An alternative is to have a `PaymentMethodAccounting.MinimumTotalDue` and use it in place of of `TotalDue` everywhere in the Watcher! Note that this make everything stupid to review!

**Comment by NicolasDorier** [comment]
I proposed some changes. But I can probably implement it myself after merge if it feel burdensome for you.

Though can you write a small test about this new feature?

**Comment by Kukks** [comment]
I can do the changes and test tomorrow unless you can't wait, no worries :)

**Comment by Kukks** [comment]
Added a very basic test and the MinimumTotaDue approach instead of IsPaid @NicolasDorier

File: BTCPayServer/HostedServices/InvoiceWatcher.cs [hunk 3]
```diff
@@ -78,13 +78,13 @@ public void MarkDirty()
...
             {
-                if (accounting.Paid >= accounting.TotalDue)
+                if (accounting.Paid >= accounting.MinimumTotalDue)
                 {
                     if (invoice.Status == "new")
                     {
                         context.Events.Add(new InvoiceEvent(invoice, 1003, "invoice_paidInFull"));
                         invoice.Status = "paid";
-                        invoice.ExceptionStatus = accounting.Paid > accounting.TotalDue ? "paidOver" : null;
+                        invoice.ExceptionStatus = accounting.Paid > accounting.MinimumTotalDue ? "paidOver" : null;
```
Comment (NicolasDorier) [inline BTCPayServer/HostedServices/InvoiceWatcher.cs]: this should be an exception where `TotalDue` is used. Else, a merchant will see the majority of people as `paidOver`

File: BTCPayServer/Services/Invoices/InvoiceEntity.cs [hunk 4]
```diff
@@ -671,6 +676,10 @@ public PaymentMethodAccounting Calculate(Func<PaymentEntity, bool> paymentPredic
             accounting.Due = Money.Max(accounting.TotalDue - accounting.Paid, Money.Zero);
             accounting.DueUncapped = accounting.TotalDue - accounting.Paid;
             accounting.NetworkFee = accounting.TotalDue - totalDueNoNetworkCost;
+            accounting.MinimumTotalDue = new Money(Convert.ToInt32(Math.Ceiling(accounting.TotalDue.Satoshi -
```
Comment (NicolasDorier) [inline BTCPayServer/Services/Invoices/InvoiceEntity.cs]: should be a `ToInt64`

**Comment by NicolasDorier** [comment]
awesome merging it, I can fix the two nits I pointed out.

**Comment by NicolasDorier** [comment]
For reference 7872b3ec55607a103a378af6ba31c6cbb99f384c

**Pull request merged**